Add preflight playbook

guits opened the pull request:
Add the `cephadm-preflight.yml` playbook.

This playbook is intended for installing prerequisites such as NTP (make sure it's installed and running), installing podman, lvm2, and cephadm so everything needed in order to bootstrap a Ceph cluster is available.

Closes: https://bugzilla.redhat.com/show_bug.cgi?id=1958675

guits committed:
tests: add functional testing

This adds some functional testing.

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

guits committed:
preflight: add playbook initial playbook

Add preflight playbook in order to install everything needed for deploying
a Ceph cluster with cephadm.

Closes: https://bugzilla.redhat.com/show_bug.cgi?id=1958675

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

guits committed:
ansible: add ansible.cfg

This adds an ansible.cfg file with some basic options

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

guits committed:
rpm: add specfile

This commit adds the needed specfile to build cephadm-ansible as an RPM

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

sebastian-philipp commented on tests/functional/Vagrantfile:
```diff
@@ -0,0 +1,145 @@
+# -*- mode: ruby -*-
+# vi: set ft=ruby :
+
+NMONS = 3
+NOSDS = 3
+NMDSS = 0
+NRGWS = 0
+NNFSS = 0
+MGRS  = 0
```
Eventually would probably make sense to not differentiate between types of hosts and only have a single type of host. It's a copy of https://github.com/ceph/ceph-ansible/blob/83a8dd5a6a1f9ffe43b2a75d7b49775e34c58f24/Vagrantfile right? I think we can simplify this later on.

guits replied:
@sebastian-philipp this file is for CI usage only, so yes we can update it later if needed

guits replied:
@sebastian-philipp, it might make sense to differentiate between types of hosts but this is another topic that has nothing to do here, I guess we should discuss it when the time comes later if we decide to have more automation on top of cephadm.

jcmdln replied:
What is the goal of splitting services onto multiple hosts in this case? Work is already being done to test that services may be deployed to individual instances by Ceph, so this seems like a duplication of effort. This playbook would likely be validated by using 3 static instances since all non-playbook reasons for that failing would be upstream bugs.

guits commented on cephadm-preflight.yml:
```diff
@@ -0,0 +1,89 @@
+- hosts: all
+  become: true
+  gather_facts: true
+  vars:
+    ceph_origin: community
+    ceph_rhcs_version: 5
+    ceph_mirror: https://download.ceph.com
+    ceph_stable_key: https://download.ceph.com/keys/release.asc
+    ceph_release: pacific
+    upgrade_ceph_packages: false
+    ceph_pkgs:
+      - lvm2
+      - chrony
+      - cephadm
+      - podman
+      - python3
```
```suggestion
```

need to set all defaults in a role

jcmdln replied:
In this case, these variables would need to be defined in `group_vars/all.yml` and not a role.

guits replied:
I don't think having default values in a versioned file intended to be modified by users is very elegant.

jcmdln replied:
Ah thank you for clarifying. Sorry I misunderstood what you meant by `role` and made an (improper) assumption that you intended to set these variables in `./roles/<somewhere>` which would violate variable precedence from when this playbook is called. 

If the defaults are to remain in this file I agree that makes a lot of sense. I've had debugging pain points with how variables are shuffled around in ceph-ansible, though they had good reasons to do so.

guits replied:
They would be defined in `./roles/ceph-defaults/defaults/main.yml`, yes.
This way you can override them in several places: inventory, group_vars/host_vars... and extra vars, if we keep them defined at the play vars level, it won't be possible (except at the extra vars level since it has the greatest precedence)

guits commented on cephadm-preflight.yml:
```diff
@@ -0,0 +1,89 @@
+      - podman
+      - python3
+  tasks:
+    - name: rhcs related tasks
```
need to support setting up dev repos

ktdreyer commented on cephadm-ansible.spec.in:
```diff
@@ -0,0 +1,42 @@
+Name:           cephadm-ansible
+Version:        @VERSION@
+Release:        @RELEASE@%{?dist}
+Summary:        Ansible playbooks for Ceph
```
Want to change this Summary?

ktdreyer commented on cephadm-ansible.spec.in:
```diff
@@ -0,0 +1,42 @@
+Version:        @VERSION@
+Release:        @RELEASE@%{?dist}
+Summary:        Ansible playbooks for Ceph
+License:        ASL 2.0 and GPLv3+
```
What is the license for this project?

ktdreyer replied:
The main ceph-ansible project is licensed as `ASL 2.0`, but there is a comment in `ceph-ansible.spec.in`:

```
# Some files have been copied from Ansible (GPLv3+). For example:
#  library/ceph_facts
#  plugins/actions/config_template.py
#  roles/ceph-common/plugins/actions/config_template.py
```

So if we're copying files from Ansible into `cephadm-ansible`, we need to indicate `GPLv3+`, but if we're only copying files from `ceph-ansible` or writing new files, we can go with `ASL 2.0`.

guits replied:
@ktdreyer I don't think we are going to copy files from Ansible, can we leave ASL 2.0 *and* GPLv3+ anyway?

ktdreyer replied:
Let's just keep it ASL 2.0 then for simplicity.

https://docs.fedoraproject.org/en-US/packaging-guidelines/LicensingGuidelines/#_dual_licensing_scenarios explains more about Dual Licensing and Multiple Licensing.

ktdreyer commented on cephadm-ansible.spec.in:
```diff
@@ -0,0 +1,42 @@
+Requires: ansible >= 2.10
+
+%description
+Ansible playbooks for Ceph
```
Let's change this description to match whatever goes into the README

ktdreyer replied:
README update in #2 says "cephadm-ansible is a collection of Ansible playbooks to simplify workflows that are not covered by cephadm."

ktdreyer commented on cephadm-ansible.spec.in:
```diff
@@ -0,0 +1,42 @@
+
+%files
+%doc README.rst
+%license LICENSE
```
Let's include a `LICENSE` file in Git for this project

jcmdln commented on cephadm-preflight.yml:
```diff
@@ -0,0 +1,89 @@
+
+
+- hosts: all
+  become: true
```
When `become` is explicitly defined this might work perfectly for users of CentOS setup via cloud-init who are implicitly given a non-root user, though when kickstarting bare-metal this likely won't be the case. Ansible already provides `--become*` flags for situations like this which could be documented above.

guits replied:
Except if you can ssh nodes as `root` (which is not something you want to do IMO), is there a scenario where you could run this playbook without `become: true` or `--become`? Could you describe your use case?
Personally, I don't mind asking people to run with `--become` but in that case, why not explicitly defining it at the play level?

jcmdln replied:
> Except if you can ssh nodes as `root` (which is not something you want to do IMO)

I am (unfortunately) that type of person :smile: 

> Personally, I don't mind asking people to run with --become but in that case, why not explicitly defining it at the play level?

After thinking about this some more, I suppose that even if _someone_ (ie me) wanted to use `root`, the use of `become` has no tangible impact. I haven't personally read over how Ansible handles that internally, though leaving `become` present skips a parameter for non-`root` users... I think `become` is also making testing with Vagrant a bit easier.

Good points though, thanks for clarifying!

guits committed:
defaults: add ceph-defaults role

This role will handle all default values that will be defined for
cephadm-ansible playbooks

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

guits committed:
preflight: support setting up dev repos

So we can test with build from ceph@master.

(mainly for CI purposes)

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

guits committed:
LICENSE: Add ASL 2.0 licence file

This adds the ASL 2.0 LICENSE file to the project.

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

Daniel-Pivonka commented on cephadm-preflight.yml:
```diff
@@ -0,0 +1,106 @@
+---
+# This playbook configures the Ceph repository.
+# It also installs some prerequisites (podman, lvm2, chronyd, ...)
```
its also is installing the cephadm binary

ktdreyer commented on LICENSE:
```diff
@@ -0,0 +1,201 @@
+      same "printed page" as the copyright notice for easier
+      identification within third-party archives.
+
+   Copyright [2014] [S�bastien Han]
```
Let's change this Copyright

ktdreyer commented on ceph-defaults/defaults/main.yml:
```diff
@@ -0,0 +1,40 @@
+ceph_pkgs:
+  - lvm2
+  - chrony
+  - cephadm
```
Does `yum install cephadm` install these others automatically? Can we reduce the duplication here?

jcmdln commented on cephadm-preflight.yml:
```diff
@@ -19,20 +19,10 @@
-      - python3
   tasks:
+    - import_role:
+        name: ceph-defaults
```
Why is the `ceph-defaults` role not in a `roles` folder?

guits replied:
could be in a roles folder or not, it doesn't really matter at the moment.
I don't mind moving it to a roles folder, in fact, perhaps it's better to keep the default behavior indeed

ktdreyer commented on tests/requirements.txt:
```diff
@@ -0,0 +1,7 @@
+pytest>=4.6,<5.0
+ansible>=2.10,<2.11,!=2.9.10
+Jinja2>=2.10
+jmespath
```
We only need `jmespath` if we were using the `json_query` filter in Ansible, and we're not using it here, so we can drop this.

Daniel-Pivonka commented on ceph-defaults/README.md:
```diff
@@ -0,0 +1,3 @@
+# Ansible role: ceph-defaults
+
+Documentation is available at http://docs.ceph.com/cephadm-ansible/.
```
i assume this link will actually go to some documentation in the future but for now should this dead link be included?

guits replied:
Yeh, just tried to anticipate when the doc will be built/available. Not a big issue IMO, we can leave it as it is for now and update it later if needed I guess.
